File: app/src/main/java/com/baozi/treerecyclerview/base/BaseRecyclerAdapter.java
package com.baozi.treerecyclerview.base;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.baozi.treerecyclerview.manager.ItemManageImpl;
import com.baozi.treerecyclerview.manager.ItemManager;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by zhy on 16/4/9.
 */
public abstract class BaseRecyclerAdapter<T> extends RecyclerView.Adapter<ViewHolder> {

    protected ItemManager<T> mItemManager;
    protected OnItemClickListener mOnItemClickListener;
    protected OnItemLongClickListener mOnItemLongClickListener;
    private List<T> mDatas;

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ViewHolder holder = ViewHolder.createViewHolder(parent, viewType);
        onBindViewHolderClick(holder, holder.itemView);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        onBindViewHolder(holder, getDatas().get(position), position);
    }

    /**
     * 实现item的点击事件
     */
    public void onBindViewHolderClick(@NonNull final ViewHolder viewHolder, View view) {
        //判断当前holder是否已经设置了点击事件
        if (!view.hasOnClickListeners()) {
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    //获得holder的position
                    int layoutPosition = viewHolder.getLayoutPosition();
                    //检查item的position,是否可以点击.
//                    检查并得到真实的position
                    int itemPosition = checkPosition(layoutPosition);
                    if (mOnItemClickListener != null) {
                        mOnItemClickListener.onItemClick(viewHolder, layoutPosition);
                    }
                }
            });
        }
        if (!view.isLongClickable()) {
            view.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    //获得holder的position
                    int layoutPosition = viewHolder.getLayoutPosition();
                    //检查position是否可以点击
                    //检查并得到真实的position
                    int itemPosition = checkPosition(layoutPosition);
                    if (mOnItemLongClickListener != null) {
                        return mOnItemLongClickListener.onItemLongClick(viewHolder, layoutPosition);
                    }
                    return false;
                }
            });
        }
    }

    @Override
    public int getItemViewType(int position) {
        return getLayoutId(position);
    }

    @Override
    public int getItemCount() {
        return getDatas().size();
    }

    public int getItemSpanSize(int position) {
        return 0;
    }

    public List<T> getDatas() {
        if (mDatas == null) {
            mDatas = new ArrayList<>();
        }
        return mDatas;
    }

    public void setDatas(List<T> datas) {
        if (datas != null && !datas.isEmpty()) {
            getDatas().clear();
            getDatas().addAll(datas);
        }
    }

    @Nullable
    public T getData(int position) {
        if (position >= 0) {
            return getDatas().get(position);
        }
        return null;
    }

    /**
     * 操作adapter
     *
     * @return
     */
    public ItemManager<T> getItemManager() {
        if (mItemManager == null) {
            mItemManager = new ItemManageImpl<T>(this);
        }
        return mItemManager;
    }

    public void setItemManager(ItemManager<T> itemManager) {
        mItemManager = itemManager;
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        mOnItemClickListener = onItemClickListener;
    }

    public void setOnItemLongClickListener(OnItemLongClickListener onItemLongClickListener) {
        mOnItemLongClickListener = onItemLongClickListener;
    }


    public interface OnItemClickListener {
        void onItemClick(@NonNull ViewHolder viewHolder, int position);
    }

    public interface OnItemLongClickListener {
        boolean onItemLongClick(@NonNull ViewHolder viewHolder, int position);
    }

//    /**
//     * 检查item属性
//     */
//    public interface CheckItemInterface {
//        int checkPosition(int position);
//    }
//
//    public void removeCheckItemInterfaces(CheckItemInterface itemInterface) {
//        if (checkItemInterfaces == null) return;
//        checkItemInterfaces.remove(itemInterface);
//    }
//
//    public void addCheckItemInterfaces(CheckItemInterface itemInterface) {
//        if (checkItemInterfaces == null) {
//            checkItemInterfaces = new ArrayList<>();
//        }
//        checkItemInterfaces.add(itemInterface);
//    }

//    private ArrayList<CheckItemInterface> checkItemInterfaces;

    //检查当前position,获取原始角标
    public int checkPosition(int position) {
//        if (checkItemInterfaces != null) {
//            for (CheckItemInterface itemInterface : checkItemInterfaces) {
//                position = itemInterface.checkPosition(position);
//            }
//        }
        return position;
    }


    /**
     * 获取该position的item的layout
     *
     * @param position 角标
     * @return item的layout id
     */
    public abstract int getLayoutId(int position);

    /**
     * view与数据绑定
     *
     * @param holder
     * @param t
     * @param position
     */
    public void onBindViewHolder(@NonNull ViewHolder holder, T t, int position) {

    }

    public void clear() {
        getDatas().clear();
    }
}

File: app/src/main/java/com/baozi/treerecyclerview/base/BaseRecyclerAdapter.kt
package com.baozi.treerecyclerview.base

import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup

import com.baozi.treerecyclerview.manager.ItemManageImpl
import com.baozi.treerecyclerview.manager.ItemManager

import java.util.ArrayList
import kotlin.properties.Delegates

/**
 * Created by zhy on 16/4/9.
 */
abstract class BaseRecyclerAdapter<T> : RecyclerView.Adapter<ViewHolder>() {
    protected var mOnItemClickListener: ((viewHolder: ViewHolder, position: Int) -> Unit)? = null
    protected var mOnItemLongClickListener: ((viewHolder: ViewHolder, position: Int) -> Boolean)? = null
    private val data: MutableList<T> = ArrayList()
    /**
     * 操作adapter
     *
     * @return
     */
    open val itemManager: ItemManager<T> by lazy {
        ItemManageImpl(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val holder = ViewHolder.createViewHolder(parent, viewType)
        onBindViewHolderClick(holder, holder.itemView)
        return holder
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        onBindViewHolder(holder, getData(position), position)
    }

    /**
     * 实现item的点击事件
     */
    open fun onBindViewHolderClick(viewHolder: ViewHolder, view: View) {
        //判断当前holder是否已经设置了点击事件
        if (!view.hasOnClickListeners()) {
            view.setOnClickListener {
                //获得holder的position
                val layoutPosition = viewHolder.layoutPosition
                //检查item的position,是否可以点击.
                //                  检查并得到真实的position
                val itemPosition = itemManager.itemToDataPosition(layoutPosition)
                if (mOnItemClickListener != null) {
                    mOnItemClickListener!!.invoke(viewHolder, itemPosition)
                }
            }
        }
        if (!view.isLongClickable) {
            view.setOnLongClickListener {
                //获得holder的position
                val layoutPosition = viewHolder.layoutPosition
                //检查position是否可以点击
                //检查并得到真实的position
                val itemPosition = itemManager.itemToDataPosition(layoutPosition)
                mOnItemLongClickListener?.invoke(viewHolder, itemPosition) ?: false
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return getLayoutId(position)
    }

    override fun getItemCount(): Int {
        return getData().size
    }


    open fun getItemSpanSize(position: Int, maxSpan: Int): Int {
        return maxSpan
    }

    open fun getData(): MutableList<T> {
        return data
    }

    open fun setData(data: List<T>?) {
        if (data != null) {
            getData().clear()
            getData().addAll(data)
        }
    }

    open fun getData(position: Int): T? {
        return if (position >= 0) {
            getData()[position]
        } else null
    }

    /**
     * 直接强转为指定类型,可以能null
     *
     * @param position
     * @param <D>
     * @return
    </D> */
    fun <D> getCastData(position: Int): D? {
        try {
            return getData(position) as D?
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }

    }

    open fun setOnItemClickListener(onItemClickListener: (viewHolder: ViewHolder, position: Int) -> Unit) {
        mOnItemClickListener = onItemClickListener
    }

    open fun setOnItemLongClickListener(onItemLongClickListener: (viewHolder: ViewHolder, position: Int) -> Boolean) {
        mOnItemLongClickListener = onItemLongClickListener
    }


    //检查当前position,获取原始角标
    open fun checkPosition(position: Int): Int {
        return itemManager.itemToDataPosition(position)
    }


    /**
     * 获取该position的item的layout
     *
     * @param position 角标
     * @return item的layout id
     */
    abstract fun getLayoutId(position: Int): Int

    /**
     * view与数据绑定
     *
     * @param holder   ViewHolder
     * @param t        数据类型
     * @param position position
     */
    open fun onBindViewHolder(holder: ViewHolder, t: T?, position: Int) {

    }

    open fun clear() {
        getData().clear()
    }
}
